Support for question references in native query template tags

walterl opened the pull request:
Fixes #11804.

* [x] Lookup native query for card with ID referenced in the `:card` template tag.
* [x] Replace card query in referencing/outer query.
* [x] Ensure that API accepts and return expected results for queries containing `:card` template tags.
* [x] Get front-end to query API with `:card` template tags.
* [x] ~Support alias for referenced card's query.~ _Alias specified in query._
* [x] Check user's permissions on referenced card.
* [x] Add support for referencing structured/MBQL cards.
* [x] ~Ensure that all of this works on non-SQL databases.~ _Leave for next round._
* [x] Implement suggested UI for the question template tags.
* [x] Ensure that sub-queries are from the same db as referencing query.
* [x] Test for circular references in recursive sub-queries.

walterl committed:
Port tests from expectations to clojure.test

walterl committed:
Fix docstring

walterl committed:
Added :question template tag type with lookup

sbelak commented on src/metabase/driver/common/parameters/values.clj:
```diff
@@ -24,6 +26,7 @@
 (def ^:private ParamType
   (s/enum :number
           :dimension                    ; Field Filter
+          :question
```
while consistent with our external nomenclature, we use card elsewhere in the code. Not sure if it's a good idea to use both terms interchangeably.

walterl replied:
Good catch. I'm unsure about where the boundary is where a "question" becomes a "card", so I opted for the front-end-friendliest option.

From your comment it seems like that boundary is where the front-end generates queries: before that it's a "question", but in the query it's a "car". Is that correct?

sbelak replied:
I think throughout the BE it's card (even the endpoint is /card)

tlrobinson replied:
> even the endpoint is /card

That's only really because we haven't taken the time to update the API.

I guess it's not that big of a deal, but I don't think we shouldn't be digging ourselves deeper by adding `card` to the API surface area, which we'll probably eventually want to migrate to `question`. I believe the only places we use `card` in the API are the `/api/card` endpoints (which could/should trivially be renamed to `/api/question` IMHO), and in dashboards.

We do use `question` in JWT embedding, for example(but that is explicitly a user-facing API)

sbelak commented on src/metabase/driver/common/parameters/values.clj:
```diff
@@ -43,14 +46,15 @@
     :display-name                 su/NonBlankString
     :type                         ParamType
     (s/optional-key :dimension)   [s/Any]
+    (s/optional-key :question)    s/Int
```
su/IntGreaterThanZero

sbelak commented on src/metabase/driver/common/parameters/values.clj:
```diff
@@ -116,6 +120,12 @@
+  "Returns the native query for the `:question` referenced by the given tag."
+  [tag :- TagParam, _params :- (s/maybe [i/ParamValue])]
+  (when-let [card-id (:question tag)]
+    (db/select-one-field :dataset_query Card :id card-id)))
```
Why is this assumed to be a native question?

walterl replied:
Mostly just to get it working. I'll keep this comment unresolved to remind me that non-native questions should be converted to native, around here. Thanks! :+1:

walterl committed:
s/question/card/

walterl committed:
Fix typo: s/substitue/substitute/g

walterl committed:
Wrap card query parameter data in a record

Just like `FieldFilter`s.

walterl committed:
Unrelated: Improve test layout

walterl committed:
Add substitution of card query parameters in queries

This implementation is incomplete.

walterl committed:
Finish substitution of card query in parent query

walterl committed:
Basic FE support for question template tags

walterl committed:
Update card ID on tag name update

Also, don't use `\d` in regexes.

walterl committed:
yarn prettier

walterl committed:
Add support for MBQL queries in template tag lookup

walterl committed:
Return query strings directly as tag value

walterl committed:
Fix query type lookup

walterl committed:
Initialize QP store for MBQL→native conversion

walterl committed:
Fix MBQL to native query conversion

mazameli commented:
@tlrobinson already covered the main UI points I noticed but I thought I'd chime in with a few ideas. My bias is the bottom mockup.

1) This is the most bare-bones thing we could do. I think the blue "Question #374" text could be a hyperlink to the question:

![image](https://user-images.githubusercontent.com/2223916/73987418-331dcd80-48f5-11ea-8545-291aa3cd9558.png)

2) Here's if we use `<QuestionPicker>`:

![image](https://user-images.githubusercontent.com/2223916/73987607-c8b95d00-48f5-11ea-9038-910122f84812.png)

3) And here's a thought about including some metadata on the selected question, like the collection it's saved in, when it was last edited, and how many rows are in the result:

![image](https://user-images.githubusercontent.com/2223916/73987696-0322fa00-48f6-11ea-942c-27607e424dbc.png)

walterl committed:
Port tests from expectations to clojure.test

walterl committed:
Add middleware to resolve tables/fields from referenced cards

This needs to happen before `substitute-parameters` in the QP pipeline,
so that `substitute-parameters` can convert any MBQL queries (from
referenced cards) into native queries. Maybe it should be consolidated
with the other resolution steps.

walterl committed:
Use mbql-query util instead of manual query def

walterl committed:
Add docstring to now public function

walterl committed:
Remove unused import

walterl committed:
Fix docstring positioning :facepalm:

walterl committed:
Clean ns declaration

walterl committed:
Convert tests to clojure.test

walterl committed:
Be more specific about errors we're looking expecting

Invalid queries caused unexpected exceptions to be thrown when checking
permissions.

walterl committed:
Recursively check permissions of referenced queries

walterl committed:
Add incomplete test [WIP]

walterl committed:
Finish permission check test for referenced MBQL queries

walterl committed:
Add tests for referenced native queries perms check

sbelak commented on src/metabase/driver/common/parameters/values.clj:
```diff
@@ -116,6 +121,17 @@
+    (when-let [query (db/select-one-field :dataset_query Card :id card-id)]
+      (i/map->CardQuery
+       {:card-id card-id
+        :query   (condp = (name (:type query))
```
why going through `name` rather than comparing keywords?

walterl replied:
Is [this](https://github.com/metabase/metabase/pull/11835/files#diff-5b28fb9bc94ad945da73a4fad99fc434R131) what you had in mind?

sbelak replied:
Yes

walterl committed:
Clean ns

walterl committed:
Use existing mbql-to-native middleware

Rather than calling driver-specific `mbql->native` directly.

walterl committed:
Fix formatting

walterl committed:
Use more comprehensive qp func for converting query to native

walterl commented:
@camsaul @sbelak While @paulrosenzweig is continuing with the front-end, I believe the back-end code is ready for review. Could either of you please have a look at it?

paulrosenzweig committed:
front end updates

paulrosenzweig commented:
## Questions

### Do we need the metadata?

![image](https://user-images.githubusercontent.com/691495/74275011-1c0a2180-4ce1-11ea-9a20-b000107c6f19.png)


I omitted it in the first pass, but could add it now if we want it for v1. I think we'd need to do another query to get the row count. Also, I don't think we should include database since they will all match.

### Should we display the picker before typing a number?

Currently, we don't add the template tag until at least one digit is typed `{{#1...`. I imagine many users know the question they want only by name. They could type "1" as a fake number to display the picker, but maybe we should pop up the picker right after typing `{{#`.

## Error states
When the question ID doesn't exist, I just left the default UI from the entity loader. That would need to change if we want to use the picker without a starting id.
![image](https://user-images.githubusercontent.com/691495/74274248-c8e39f00-4cdf-11ea-9b05-874a76b8f511.png)
If the question does exist but the database doesn't match, I added a note.
![image](https://user-images.githubusercontent.com/691495/74274293-e0228c80-4cdf-11ea-927b-dc584069dac4.png)

walterl commented:
@paulrosenzweig 

> Do we need the metadata?

Barring design considerations, I would opt for any metadata that would help make it clear that the selected query is the one that I (as a user) was looking for. @mazameli?

> Should we display the picker before typing a number?

IMHO, yes. By the time a user types the `#` in `{{#`, they're already unambiguously creating a query template tag, and we can therefore go ahead and display the query picker.

mazameli commented:
> Do we need the metadata?

I actually don't think we need the row count if that's a hassle. The name of the collection the question is stored in would be useful (that's what the `Baseball` item in the mock is meant to represent; not database), as would be the last-edited-at date if that's already accessible in the question's metadata. Basically I'm trying to make it easier to verify that yes, this is the question I meant to select, and make edits to referenced questions more transparent to help with debugging e.g. if someone accidentally moved something out from under you.

> Should we display the picker before typing a number?

I agree with what @walterl said. 👍 

> If the question does exist but the database doesn't match, I added a note.

Blerg, yeah, I guess this is necessary. I was going to say I'd prefer it if the picker would only list saved questions based on the same database as the currently selected one, but then you still have the scenario where someone types in the number of a question from a different DB. I'd suggest the error text to read, "This question can't be used because it's based on a different database."

> When the question ID doesn't exist, I just left the default UI from the entity loader. That would need to change if we want to use the picker without a starting id.

Let me quickly mock this up with the picker on the screen as well.

walterl commented:
@mazameli 

> I was going to say I'd prefer it if the picker would only list saved questions based on the same database as the currently selected one, but then you still have the scenario where someone types in the number of a question from a different DB.

Can't we have both? That is, only list questions from the same db in the picker, but load questions and display error message for questions from other db's.

mazameli commented:
@walterl yeah, you're probably right. I was just getting hung up on the case where you type in an ID number referencing a question from a different DB and assuming that the picker would need to display the title of that question for clarity, but I guess it doesn't have to.

mazameli commented:
Okay, here are some mocks for the states @paulrosenzweig pointed out.

### Hash character typed, but no numbers

![image](https://user-images.githubusercontent.com/2223916/74282077-60db8b80-4cd4-11ea-95d1-27e8df7cd859.png)

### Question ID doesn't exist

![image](https://user-images.githubusercontent.com/2223916/74282136-7f418700-4cd4-11ea-9d79-d1824d089dc2.png)

### Reference a question from a different database

![image](https://user-images.githubusercontent.com/2223916/74282277-af892580-4cd4-11ea-9791-8a68c38d0c85.png)

walterl committed:
Check that all referenced queries are from the same db

walterl committed:
Add test for checking db id of referenced MBQL query

walterl committed:
Clean ns

sbelak reviewed (changes requested):
It would be good to have tests covering referencing multiple queries (eg. multiple WITH AS statements) and nesting 2+ levels deep.

sbelak commented on src/metabase/query_processor/middleware/resolve_referenced.clj:
```diff
@@ -0,0 +1,36 @@
+  (->> (get-in query [:native :template-tags])
+       vals
+       (filter #(= (keyword (:type %)) :card))
+       (map :card)
```
since only type=card has a `:card` field, vals+filter+map could be simplified to `(keep (comp :card val)`

paulrosenzweig committed:
add metadata and better error states to UI

paulrosenzweig committed:
Merge branch 'master' into 11804-reference-question-as-cte

paulrosenzweig commented:
I implemented the mocks. One thing to note is the behavior for cards in the root collection. I hid that part of the metadata rather than calling it "Root Collection".

![image](https://user-images.githubusercontent.com/691495/74353516-04d23f00-4d88-11ea-93c3-5ca2fd553960.png)

![image](https://user-images.githubusercontent.com/691495/74353497-fedc5e00-4d87-11ea-8c5a-04a263e1aef7.png)

---

I didn't update the question picker to only show cards from the relevant database since database_id isn't returned from that search endpoint.

walterl commented on src/metabase/driver/common/parameters.clj:
```diff
@@ -27,6 +27,21 @@
+  (pretty [this]
+    (list 'map->CardQuery (into {} this))))
+
+(defn CardQuery?
```
I really like that, but it doesn't quite fit: this structure holds _primarily_ the query of the referenced card, and the card ID as metadata.

I'm not too fond of `CardQuery` either, but the best I could think of. Is `ReferencedCardQuery` too long?

sbelak replied:
The longer I'm at MB, the more I'm comfortable with very long but also very descriptive names.

camsaul replied:
I sort of picked this up from back when I used to do iOS apps where you have methods calls like

```objc
[NSLayoutConstraint constraintWithItem:loadingView
                    attribute:NSLayoutAttributeWidth
                    relatedBy:NSLayoutRelationEqual
                    toItem:nil       
                    attribute:NSLayoutAttributeNotAnAttribute
                    multiplier:1.0f
                    constant:200.0f]
```

very verbose but I would also argue very readable and hard to get the params wrong.

This sort of explains the philosophy: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingBasics.html#//apple_ref/doc/uid/20001281-1001751-BBCFECGB

I don't think going as far as Objective-C is necessary but more verbose and unambiguous is a good thing IMO.

walterl replied:
Yeah, 3 words is probably well below the "too long" threshold. Thanks for that link.

walterl committed:
Combine `vals`+`filter`+`map` into a single `keep`

paulrosenzweig committed:
move error message beneath the question picker

walterl committed:
Rename `CardQuery`→`ReferencedCardQuery`

walterl committed:
Cosmetic: fix typo and formatting

walterl committed:
Test substitution of multiple sub-queries

walterl committed:
Test CTE syntax substitution of multiple sub-queries

walterl commented:
@sbelak 

> It would be good to have tests covering referencing multiple queries (eg. multiple WITH AS statements)

[Done](https://github.com/metabase/metabase/pull/11835/files#diff-d3897fca8ef0d4310470092491f06c72R142).

> ... and nesting 2+ levels deep.

We are not looking to support multiple levels of substitution with this implementation.

walterl committed:
Clean ns

salsakran commented:
@walterl "We are not looking to support multiple levels of substitution with this implementation"
where did that come from?

How will users know if a question they include in a variable has a sub-question?

walterl commented:
@salsakran 

> where did that come from?

When we discussed inherited template tags, didn't you mention that we're just looking to support a simple "library of common sub-queries" (paraphrasing from memory)?

If I misunderstood, we'll also need to check for circular references, which I've also ignored until now.

walterl committed:
Clean ns

sbelak commented:
I think these are 2 separate cases. One is referencing questions with what we call variables; the other is referencing questions that reference further questions. Punting on the first makes sense and doesn't overly detract from the usability of this, but I think we should do the latter. 

p.s. if you need cycle detection, we already have one in https://github.com/metabase/metabase/blob/master/src/metabase/util.clj#L685

salsakran commented:
So we'd need to do one of three things
1) Not allow the user to select a question that itself has a sub-question
2) Throw a warning if a user did that
3) Actually insert saved question template variables properly regardless of nesting

I think (3) is the "Metabasey" way of doing things. Aka, use recursion and do it right the first time rather than introducing weird edge cases and errors to avoid implementing things properly the first time.

walterl committed:
Simplify referenced card substitution tests

walterl committed:
Test recursive sub-query substitution

walterl committed:
Cosmetic: typos

walterl committed:
Fix parameter names in comments

walterl committed:
Check that queries don't include circular sub-queries

walterl committed:
Clean ns

walterl commented:
@sbelak I managed to get a recursive, `reduce`-based [circular dependency checking solution](https://github.com/metabase/metabase/pull/11835/commits/a46abad075e726e111b57789e5cc9c6bbd2a994a) going, but didn't manage to figure out how to make that work with `transduce`. Suggestions (and review) welcome. :slightly_smiling_face:

walterl committed:
Give users a more user-friendly error for circular referencing sub-queries

walterl committed:
Test referencing queries with parameters

walterl commented:
@mazameli @kdoh  Could either of you please check if you're happy with the copy of [this error message](https://github.com/metabase/metabase/pull/11835/commits/53dce8b6d9f3dd1c7967263c1f81219478d5ea2f#diff-9ba9c64872dacaa325097df8bcbca08bR58)? It is displayed when there is a circular reference somewhere in the graph of queries referencing each other, as sub-queries. The circular reference is detected when registering the sub-query link between the two named queries.

mazameli commented on src/metabase/query_processor/middleware/resolve_referenced.clj:
```diff
@@ -50,10 +52,23 @@
+(defn- circular-ref-error
+  [from-card to-card]
+  (tru
+   "Your query includes circular referencing sub-queries between cards \"{0}\" and \"{1}\"."
```
```suggestion
   "This query includes saved question #{0} and #{1}, but they can't reference each other."
```

mazameli replied:
Okay, how 'bout this:

```suggestion
   "This query has circular referencing sub-queries. These questions seem to be part of the problem: "\"{0}\" and \"{1}\"."
```

walterl committed:
Update src/metabase/query_processor/middleware/resolve_referenced.clj

Co-Authored-By: Maz Ameli <maz@metabase.com>

walterl committed:
Fix error message and combine card name queries

sbelak commented on src/metabase/query_processor/middleware/resolve_referenced.clj:
```diff
@@ -28,7 +30,33 @@
+  (keep :card (query->template-tags query)))
+
+(defn- card-subquery-graph
+  ([card-id]
```
Where is this airity used?

sbelak reviewed (approved):
This looks fine. I don't think anything would be gained by rewriting this to use transducers.

walterl commented:
Thanks, @sbelak!

Upon testing this a bit further, I'm looking into improving the error handling of sub-queries with required parameters. At the moment it just responds with a generic "query failed" message.

walterl committed:
Wrap sub-query error in a more user friendly explanation

walterl committed:
Cosmetic: function argument alignment

walterl committed:
Test error handling of referenced sub-queries

walterl committed:
Remove unused 1-arity version of `card-subquery-graph`

walterl commented:
@sbelak I updated error handling. Could you maybe just have one last look at [this commit](https://github.com/metabase/metabase/pull/11835/commits/ca98852c7cb255e91dcb86c6ce0f5a3bbf7fa443) (with [test](https://github.com/metabase/metabase/pull/11835/commits/cadd61bad35cc60b35570f1e6d43fdc731faeb13)), please?

@mazameli @kdoh Could either of you please check if you're happy with the copy of [this error message](https://github.com/metabase/metabase/pull/11835/commits/ca98852c7cb255e91dcb86c6ce0f5a3bbf7fa443#diff-5b28fb9bc94ad945da73a4fad99fc434R138)? It comes out like this when a query is referenced with a missing parameter value: 
![image](https://user-images.githubusercontent.com/23798/74691757-bb9d4780-51ec-11ea-988d-208611d1a6a1.png)

It also does the "right thing" for nested queries: 
![image](https://user-images.githubusercontent.com/23798/74691791-e091ba80-51ec-11ea-81b0-71c197fc3879.png)

While it works well, I'm not sure if the copy is up to standard.

Unlike the [circular reference error message](https://github.com/metabase/metabase/pull/11835/files#diff-9ba9c64872dacaa325097df8bcbca08bR57-R58), I opted for using the question IDs here, because the outer/first one is guaranteed to be present in the query that the user is currently writing, allowing for an easy visual match.

sbelak commented on src/metabase/driver/common/parameters/values.clj:
```diff
@@ -116,6 +122,28 @@
+  "Returns the native query for the `:card` referenced by the given tag."
+  [tag :- TagParam, _params :- (s/maybe [i/ParamValue])]
+  (when-let [card-id (:card tag)]
+    (when-let [query (db/select-one-field :dataset_query Card :id card-id)]
```
In what case would this be false, ie. we have an id, but it doesn't map to an actual card?

walterl replied:
`card-id` isn't validated anywhere else on the back-end, so it may be invalid.

sbelak replied:
Fair, but wouldn't it be then helpful to surface this error to the user, rather than silently failing?

walterl replied:
You're right.

walterl replied:
I left this as-is, but added validation [here](https://github.com/metabase/metabase/pull/11835/files#diff-9ba9c64872dacaa325097df8bcbca08bR27), because invalid card IDs trigger ["invalid database" errors](https://github.com/metabase/metabase/pull/11835/files#diff-9ba9c64872dacaa325097df8bcbca08bR42) before reaching this point.

sbelak commented on src/metabase/driver/sql/parameters/substitution.clj:
```diff
@@ -261,3 +262,11 @@
+(defmethod ->replacement-snippet-info [:sql ReferencedCardQuery]
+  [_driver {:keys [query]}]
+  {:prepared-statement-args nil
+   :replacement-snippet (str "(" query ")")})
```
indentation

sbelak commented on src/metabase/query_processor/middleware/resolve_referenced.clj:
```diff
@@ -0,0 +1,75 @@
+(defn tags-referenced-cards
+  "Returns Card instances referenced by the given native `query`."
+  [query]
+  (->> (get-in query [:native :template-tags])
```
why doesn't this use `query->template-tags` & `query->tag-card-ids`?

walterl replied:
Only because of the order I wrote these functions in. :wink:

Fixed.

sbelak reviewed (approved):
LGTM

tlrobinson commented on frontend/src/metabase-lib/lib/queries/NativeQuery.js:
```diff
@@ -44,6 +44,19 @@ export const NATIVE_QUERY_TEMPLATE: NativeDatasetQuery = {
+function cardTagCardId(name) {
+  const reCardTag = /^#([0-9]*)$/g;
+  const match = reCardTag.exec(name);
+  if (match !== null && match[1].length > 0) {
```
I'd probably do

```
const match = name.match(/^#([0-9]+)$/);
if (match) {
  return parseInt(match[1])
}
```

I avoid `exec` for the reason I think you discovered (it's not stateless)

tlrobinson replied:
Also we allow whitespace in template tags, so maybe change the regex to `/^#\s*([0-9]+)\s*$/`, move it to a `const CARD_TAG_REGEX` outside the function and use in both places (ideally in `replaceCardId` too but that's a little trickier)

tlrobinson commented on frontend/src/metabase-lib/lib/queries/NativeQuery.js:
```diff
@@ -318,13 +339,18 @@ export default class NativeQuery extends AtomicQuery {
+          newTag.name = newTags[0];
+          if (isCardQueryName(newTag.name)) {
+            newTag.type = "card";
+            newTag.card = cardTagCardId(newTag.name);
```
I'd call this `card_id`

tlrobinson commented on frontend/src/metabase/query_builder/components/template_tags/CardTagEditor.jsx:
```diff
@@ -0,0 +1,150 @@
+import { formatDateTimeWithUnit } from "metabase/lib/formatting";
+import MetabaseSettings from "metabase/lib/settings";
+
+function mapStateToProps(state, props) {
```
Unless I'm missing something, I think this could all be replaced with the question entity loader (the prop will be named `question` instead of `card`, though)

```
import Question from "metabase/entities/questions";

@Questions.load({ id: (state, { tag }) => tag.card })
```

Actually, it looks like that would conflict with the `question` prop already being passed in. I still think it would be better to figure out how to use the Question entity loader though.

One idea would be adding a prop to `EntityObjectLoader` (and `EntityListLoader`) to allow you to pick the alias which overrides the default (`entityDef.nameOne`):

```
@Questions.load({
  id: ...,
  alias: "card"
})
```

replacing

```
    return children({
      ..._.omit(props, ...CONSUMED_PROPS),
      object,
      // alias the entities name:
      [entityDef.nameOne]: object,
      reload: this.reload,
      remove: this.remove,
    });
```

with 

```
    return children({
      ..._.omit(props, ...CONSUMED_PROPS),
      object,
      // alias the entities name:
      [props.alias || entityDef.nameOne]: object,
      reload: this.reload,
      remove: this.remove,
    });
```

Or maybe a generic way of picking/renaming parameters:

```
@Questions.load({
  id: ...,
  props: ({ object }) => ({ card: object })
})
```

Either way make sure to add the new prop to `CONSUMED_PROPS`

paulrosenzweig replied:
I had started with the entity loader but I removed it to customize the 404 behavior. Though, now I remember there's a flag to suppress that anyways. I'll swap it back in.

tlrobinson commented on frontend/src/metabase/query_builder/components/template_tags/CardTagEditor.jsx:
```diff
@@ -0,0 +1,150 @@
+}
+
+// This formats a timestamp as a date using any custom formatting options.
+function formatDate(value) {
```
This should probably be in `metabase/lib/formatting` but currently that code always assumes the custom formatting is passed in via `column` settings, so maybe this is ok for now.

paulrosenzweig replied:
Yeah, it is formatting code, but it's not for question data. 🤷‍♂Are there any other places in the app where we want to use general formatting settings for "UI data"?

Also, these timestamps seem to come down without TZ data, so there's probably some issues there.

tlrobinson replied:
There are some places we use `formatNumber` (maybe others) on non-results, but yeah, I'm not sure we should be conflating the two use cases.

walterl committed:
Cosmetic: indentation

walterl committed:
Normalize l10n of error messages

walterl committed:
Reuse functions for looking up referenced cards from query

paulrosenzweig committed:
implement FE PR feedback

paulrosenzweig committed:
card -> card_id backend changes

paulrosenzweig committed:
fix tests

paulrosenzweig committed:
fix another test

walterl commented:
Paul's back-end changes were tested and LGTM :+1:

tlrobinson commented on frontend/src/metabase-lib/lib/queries/NativeQuery.js:
```diff
@@ -44,17 +44,19 @@ export const NATIVE_QUERY_TEMPLATE: NativeDatasetQuery = {
 };
 
+// This regex needs to match logic in replaceCardId and _getUpdatedTemplateTags.
+const CARD_TAG_REGEX = /^#\s*([0-9]*)\s*$/;
```
I think this should start with`/^\s*` to allow `{ #123}` (in addition to `{# 123}` and  `{#123 }`)

paulrosenzweig replied:
It does allow for that. This regex is used after [parsing the tag name](https://github.com/metabase/metabase/pull/11835/files#diff-98acf6549faf2b8c61213474d5146919R326) out of the query. Any leading space within the braces isn't captured.

paulrosenzweig replied:
That's also true for trailing space, so I guess I can drop that in this regex.

tlrobinson commented on frontend/src/metabase-lib/lib/queries/NativeQuery.js:
```diff
@@ -44,17 +44,19 @@ export const NATIVE_QUERY_TEMPLATE: NativeDatasetQuery = {
-  const match = reCardTag.exec(name);
-  if (match !== null && match[1].length > 0) {
+  const match = name.match(CARD_TAG_REGEX);
+  if (match != null && match[1].length > 0) {
```
Super minor, personally I'd leave off ` != null` but NBD.

tlrobinson commented on src/metabase/driver/common/parameters/values.clj:
```diff
@@ -48,7 +48,7 @@
     :type                         ParamType
     (s/optional-key :dimension)   [s/Any]
-    (s/optional-key :card)        su/IntGreaterThanZero
+    (s/optional-key :card-id)     su/IntGreaterThanZero
```
I see the FE uses `card_id` and the BE uses `card-id`. Is that automatically converted somewhere?

paulrosenzweig replied:
It is! Though, I'm not sure where that happens.

tlrobinson reviewed (changes requested):
One minor regex issue and question about `card-id`

paulrosenzweig committed:
disallow space between # and number

paulrosenzweig commented:
I just undid the change to allow space between the `#` and the number. It created a weird situation where two identical card ids could differ on whitespace and appear twice in the sidebar. We could normalize the template tag names to avoid this, but it seemed simpler to just constrain how they're entered.

tlrobinson commented on frontend/src/metabase-lib/lib/queries/NativeQuery.js:
```diff
@@ -45,11 +45,11 @@ export const NATIVE_QUERY_TEMPLATE: NativeDatasetQuery = {
 
 // This regex needs to match logic in replaceCardId and _getUpdatedTemplateTags.
-const CARD_TAG_REGEX = /^#\s*([0-9]*)\s*$/;
+const CARD_TAG_REGEX = /^#([0-9]*)$/;
```
We should still allow whitespace before and after `#([0-9]*)`, no? It's allowed in other template tag types, and the two other regexes below.

paulrosenzweig replied:
Yes space around `#[0-9]*` is still allowed. This regex is used after that group is extracted from [here](https://github.com/metabase/metabase/blob/11804-reference-question-as-cte/frontend/src/metabase-lib/lib/queries/NativeQuery.js#L326).

walterl replied:
Just to confirm: was the `*` this regex deliberately chosen? I.e. shouldn't we require at least one number for a match (`+`)?

paulrosenzweig replied:
We parse `{{#}}` as a card tag, but if the captured group is `""`, we set the card_id to null. That's triggers the "Pick a question" UI.

walterl committed:
Merge branch 'master' into 11804-reference-question-as-cte

walterl committed:
Align new middleware with newer QP architecture

walterl committed:
Align sub-query error handling with the new QP

It seems like exceptions are handled more consistently in the new QP, so
we can be less hacky.

walterl committed:
(Re)wrap sub-query errors in user-friendly expalanation

walterl committed:
Merge remote-tracking branch 'origin/master' into 11804-reference-question-as-cte

walterl committed:
Rename `xformf`→`rff` to align with QP changes

paulrosenzweig committed:
force cypress click on select placeholder

paulrosenzweig commented:
I fixed the failing test.

@mazameli @tlrobinson 👍from you is all that's left I think!

mazameli commented:
@paulrosenzweig Sweet. I'll spend some time looking at it this evening.

mazameli commented:
Looks like this popover is getting incorrectly positioned such that it overflows the right side of the screen and causing horizontal scrolling:

![image](https://user-images.githubusercontent.com/2223916/75205532-e164a800-5728-11ea-8136-fd983211da1b.png)

mazameli commented:
The trailing braces and the rest of the line after the question number should not be green/teal.
And the part that we _should_ colorize, the `#497` bit, we should use brand-blue instead of green, to tie it in to the sidebar color choices.

![image](https://user-images.githubusercontent.com/2223916/75206010-37861b00-572a-11ea-88a4-5246322f518f.png)

mazameli commented:
Some more question selection weirdness if the variables side panel has enough content to scroll, and if there are lots of questions in the root level collection. I can't scroll to the start of the list of subcollections to navigate to a different one.

![popover-weirdness](https://user-images.githubusercontent.com/2223916/75217344-a6c03700-574b-11ea-90ec-3fd8fecc3b5b.gif)

mazameli commented:
A minor annoyance having to do with our editor: it seems that when you try to do a _second_ CTE tag on the _same line_, when you enter two new open curly braces, for some reason it only auto inserts _one_ matching closing brace so that when you then type a hash mark, a new variable box doesn't show up.

![second-brace](https://user-images.githubusercontent.com/2223916/75217739-bc822c00-574c-11ea-8790-625c08e3a249.gif)

mazameli commented:
Bit of polish: if I enter the same question ID as another CTE variable that I already have, the variable box will disappear. Do we need to prohibit the user from selecting the same question twice? The reason I'd prefer to not have this kind of behavior is it could make that side panel feel glitchy in cases where you already have e.g. question `3` referenced, and then as you're trying to add a second one and start typing `#35`, the box will momentarily disappear since for a minute you'll have entered `{{#3}}`.

![disappearing-double](https://user-images.githubusercontent.com/2223916/75218081-f869c100-574d-11ea-9d6e-c420633f617f.gif)

paulrosenzweig committed:
move popover attachement and cap its height

paulrosenzweig commented:
@mazameli thanks for those polish items!

I fixed the popover attachment issue and the problem where the popover would go offscreen without scrolling.

---

I can't seem to reproduce either of the highlighting issues you saw. Any idea what I could be doing differently?
![image](https://user-images.githubusercontent.com/691495/75290451-5a7a0300-57ee-11ea-9894-e54fca033548.png)

![two-same-line](https://user-images.githubusercontent.com/691495/75290957-51d5fc80-57ef-11ea-879e-c5113cacc272.gif)

---

I see what you mean with the weirdness of seeing a template tag disappear when it references the same question. We could get around that, but it introduces some trickiness around replacing only the correct ID in the query. It's definitely possible, but I think we should punt on it for now.

We have the same behavior for normal template tags. Should that also change?

![repeated-variable](https://user-images.githubusercontent.com/691495/75291306-f3f5e480-57ef-11ea-851d-5db335cd0e3f.gif)

mazameli commented:
@paulrosenzweig sweet! 

- Highlighting issues: let me pull and try again. I'll also try in some other browsers (I only tested Brave).
- Disappearing tag item in sidebar: it seems edge-casey enough that if it's tricky, I'm okay with punting.

mazameli commented:
@paulrosenzweig, the reason you're not seeing the syntax highlighting I had in my screenshot is because the highlighting is different depending on the type of database you're using. 🤦‍♂ I think my weird green highlighting is only happening with MySQL.

H2:

![image](https://user-images.githubusercontent.com/2223916/75293903-06feb980-57dc-11ea-9052-4efc16c235cb.png)

MySQL:

![image](https://user-images.githubusercontent.com/2223916/75293916-11b94e80-57dc-11ea-88d3-3d9451ae4388.png)

Redshift:

![image](https://user-images.githubusercontent.com/2223916/75293971-27c70f00-57dc-11ea-8029-7cfdc41f610d.png)

Postgresql:

![image](https://user-images.githubusercontent.com/2223916/75294104-74124f00-57dc-11ea-8af8-18ea642ec80b.png)

mazameli commented:
@paulrosenzweig Oh, and I can no longer repro that problem with the matching curly brace not showing up, so it might've just been a glitch in my session or something. 🤷‍♂

mazameli commented:
Any good reason this is day/month/year instead of `Dec 19, 2019`?

![image](https://user-images.githubusercontent.com/2223916/75294441-4548a880-57dd-11ea-8c8f-85e7c108bea5.png)

paulrosenzweig commented:
@mazameli ah! that makes sense since comment syntax is dialect-specific. I'll check if there's a way to tweak that.

That date format should pull from the instance's settings.

mazameli committed:
icon spacing

mazameli reviewed (approved):
Besides the small polish items I listed, this is feeling pretty solid, and I think it's ready to merge. I've done a fair bit of real querying with it, but I'd want to see us stress test this a bit with some real use cases before the final version of 0.35 ships.

paulrosenzweig committed:
customize highlighting for template tags

paulrosenzweig commented:
I added in blue highlighting for all template tags:

![image](https://user-images.githubusercontent.com/691495/75372663-d891e480-5896-11ea-8b29-9a72e0095a96.png)
